### PR TITLE
[DLS-8416] Adding PackagedAsContractPackerPage and HowManyPackagedAsC…

### PR DESCRIPTION
**Pull request by JakeReid2020**
…ontractPackerPage

### CODE REVIEW

File: app/views/correctReturn/PackagedAsContractPackerView.scala.html [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        @govukDetails(
+            Details(
+                summary = Text(messages("correctReturn.packagedAsContractPacker.detailsLink")),
+                content = HtmlContent(htmlContent)
+            )
+        )
+
+        @govukButton(
+            ButtonViewModel(messages("site.continue"))
```
Comment (ginnyamazed) [inline app/views/correctReturn/PackagedAsContractPackerView.scala.html:64]: site.saveContinue

File: conf/messages.en [hunk 2]
```diff
@@ -317,3 +317,19 @@ correctReturn.select.help.content = If you did not send a return, then it will n
 correctReturn.select.checkYourAnswersLabel = Which return do you need to correct?
 correctReturn.select.error.required = Select which return you need to correct
 correctReturn.select.change.hidden = Select
+
+correctReturn.packagedAsContractPacker.title = Are you reporting liable drink you have packaged as a third party or contract packer at UK sites you operate?
+correctReturn.packagedAsContractPacker.heading = Are you reporting liable drink you have packaged as a third party or contract packer at UK sites you operate?
```
Comment (ginnyamazed) [inline conf/messages.en:322]: drinks

Comment (ginnyamazed) [inline conf/messages.en:322]: drinks

File: conf/messages.en [hunk 3]
```diff
@@ -317,3 +317,19 @@ correctReturn.select.help.content = If you did not send a return, then it will n
...
+correctReturn.packagedAsContractPacker.checkYourAnswersLabel =
```
Comment (ginnyamazed) [inline conf/messages.en:323]: please put in based on returns info

File: conf/messages.en [hunk 4]
```diff
@@ -317,3 +317,19 @@ correctReturn.select.help.content = If you did not send a return, then it will n
...
+correctReturn.packagedAsContractPacker.detailsLink = What is a small producer?
+correctReturn.packagedAsContractPacker.detailsContent = A business is a small producer if it:
+correctReturn.packagedAsContractPacker.bullet1 = has had less than 1 million litres of its own brands of liable drink packaged globally in the past 12 months
+correctReturn.packagedAsContractPacker.bullet2 = will not have more than 1 million litres of its own brands of liable drink packaged globally in the next 30 days
+correctReturn.packagedAsContractPacker.error.required = Select yes if are you reporting liable drink you have packaged as a third party or contract packer at UK sites you operate?
```
Comment (ginnyamazed) [inline conf/messages.en:328]: drinks

File: conf/messages.en [hunk 5]
```diff
@@ -317,3 +317,19 @@ correctReturn.select.help.content = If you did not send a return, then it will n
...
+correctReturn.packagedAsContractPacker.change.hidden = are you reporting liable drink you have packaged as a third party or contract packer at UK sites you operate?
```
Comment (ginnyamazed) [inline conf/messages.en:329]: drinks

File: it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+          setAnswers(userAnswers)
+
+          WsTestClient.withClient { client =>
+            val result1 = createClientRequestGet(client, correctReturnBaseUrl + normalRoutePath)
+
+            whenReady(result1) { res =>
+              res.status mustBe 200
+              val page = Jsoup.parse(res.body)
+              page.title must include(Messages("correctReturn.packagedAsContractPacker" + ".title"))
```
Comment (ginnyamazed) [inline it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala:59]: same as above

File: it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+        setAnswers(emptyUserAnswersForCorrectReturn)
+
+        WsTestClient.withClient { client =>
+          val result1 = createClientRequestGet(client, correctReturnBaseUrl + checkRoutePath)
+
+          whenReady(result1) { res =>
+            res.status mustBe 200
+            val page = Jsoup.parse(res.body)
+            page.title must include(Messages("correctReturn.packagedAsContractPacker" + ".title"))
```
Comment (ginnyamazed) [inline it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala:90]: same as above

File: it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+          whenReady(result) { res =>
+            res.status mustBe 400
+            val page = Jsoup.parse(res.body)
+            page.title must include("Error: " + Messages("correctReturn.packagedAsContractPacker" + ".title"))
+            val errorSummary = page.getElementsByClass("govuk-list govuk-error-summary__list")
+              .first()
+            errorSummary
+              .select("a")
+              .attr("href") mustBe "#value"
+            errorSummary.text() mustBe Messages("correctReturn.packagedAsContractPacker" + ".error.required")
```
Comment (ginnyamazed) [inline it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala:213]: please put actual error

File: it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        setAnswers(emptyUserAnswersForCorrectReturn)
+        WsTestClient.withClient { client =>
+          val result = createClientRequestPOST(
+            client, correctReturnBaseUrl + checkRoutePath, Json.obj("value" -> "")
+          )
+
+          whenReady(result) { res =>
+            res.status mustBe 400
+            val page = Jsoup.parse(res.body)
+            page.title must include("Error: " + Messages("correctReturn.packagedAsContractPacker" + ".title"))
```
Comment (ginnyamazed) [inline it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala:296]: please put actual message

File: it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+            val errorSummary = page.getElementsByClass("govuk-list govuk-error-summary__list")
+              .first()
+            errorSummary
+              .select("a")
+              .attr("href") mustBe "#value"
+            errorSummary.text() mustBe Messages("correctReturn.packagedAsContractPacker" + ".error.required")
```
Comment (ginnyamazed) [inline it/controllers/correctReturn/PackagedAsContractPackerControllerISpec.scala:302]: same as above

File: test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        val documentFormErrorsNegative = doc(htmlFormErrorsNegative)
+        val htmlFormErrorsNoneNumeric: HtmlFormat.Appendable = howManyPackagedAsContractPackerView(formWithNoNumeric, mode)
+        val documentFormErrorsNoneNumeric = doc(htmlFormErrorsNoneNumeric)
+        val htmlFormErrorsNotWhole: HtmlFormat.Appendable = howManyPackagedAsContractPackerView(formWithDecimalNumber, mode)
+        val documentFormErrorsNotWhole = doc(htmlFormErrorsNotWhole)
+        val htmlFormErrorsOutOfRange: HtmlFormat.Appendable = howManyPackagedAsContractPackerView(formWithOutOfRangeNumber, mode)
+        val documentFormErrorsOutOfRange = doc(htmlFormErrorsOutOfRange)
+
+        "should have the expected title" in {
+          document.title() must include(Messages("correctReturn.howManyPackagedAsContractPacker.title"))
```
Comment (ginnyamazed) [inline test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala:54]: please put actual message

File: test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        "should have the expected heading" in {
+          document.getElementsByClass(Selectors.heading).text() mustBe Messages("correctReturn.howManyPackagedAsContractPacker.heading")
```
Comment (ginnyamazed) [inline test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala:58]: same as above

File: test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        "should include a govuk body with the expected content" in {
+          document.getElementsByClass(Selectors.body).first().text() mustBe Messages("correctReturn.howManyPackagedAsContractPacker.subtext")
```
Comment (ginnyamazed) [inline test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala:62]: same as above

File: test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        testLitresInBandsWithPrepopulatedData(documentWithValidData)
+
+        val expectedDetails = Map(
+          Messages("correctReturn.packagedAsContractPacker.detailsLink") -> Messages ("correctReturn.packagedAsContractPacker.detailsContent") )
+        testButton(document)
+        testAction(document, routes.HowManyPackagedAsContractPackerController.onSubmit(mode).url)
+
+
+        "and the form has errors" - {
+          val errorTitle = "Error: " + Messages("correctReturn.howManyPackagedAsContractPacker.title")
```
Comment (ginnyamazed) [inline test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala:75]: same as above

File: test/views/correctReturn/PackagedAsContractPackerViewSpec.scala [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+    val errorSummaryList = "govuk-list govuk-error-summary__list"
+    val button = "govuk-button"
+    val form = "form"
+  }
+
+  "View" - {
+    val html = view(form, NormalMode)(request, messages(application))
+    val document = doc(html)
+    "should contain the expected title" in {
+      document.title() must include(Messages("correctReturn.packagedAsContractPacker" + ".title"))
```
Comment (ginnyamazed) [inline test/views/correctReturn/PackagedAsContractPackerViewSpec.scala:51]: same as above

File: test/views/correctReturn/PackagedAsContractPackerViewSpec.scala [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    "should include a legend with the expected heading" in {
+      val legend = document.getElementsByClass(Selectors.legend)
+      legend.size() mustBe 1
+      legend.get(0).getElementsByClass(Selectors.heading).text() mustEqual Messages("correctReturn.packagedAsContractPacker.heading")
```
Comment (ginnyamazed) [inline test/views/correctReturn/PackagedAsContractPackerViewSpec.scala:57]: same as above

File: test/views/correctReturn/PackagedAsContractPackerViewSpec.scala [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+        }
+      }
+    }
+
+    "when there are form errors" - {
+      val htmlWithErrors = view(form.bind(Map("value" -> "")), NormalMode)(request, messages(application))
+      val documentWithErrors = doc(htmlWithErrors)
+
+      "should have a title containing error" in {
+        val titleMessage = Messages("correctReturn.packagedAsContractPacker.title")
```
Comment (ginnyamazed) [inline test/views/correctReturn/PackagedAsContractPackerViewSpec.scala:205]: same as above

File: test/views/correctReturn/PackagedAsContractPackerViewSpec.scala [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        errorSummary.text() mustBe Messages("correctReturn.packagedAsContractPacker.error.required")
+      }
+    }
+
+    testBackLink(document)
+    validateTimeoutDialog(document)
+    validateTechnicalHelpLinkPresent(document)
+    validateAccessibilityStatementLinkPresent(document)
+  }
+
```
Comment (ginnyamazed) [inline test/views/correctReturn/PackagedAsContractPackerViewSpec.scala:225]: page should also be verifying the detail info about a small producer

File: test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+          testEmptyFormErrors(documentFormErrorsEmpty, errorTitle)
+          testNoNumericFormErrors(documentFormErrorsNoneNumeric, errorTitle)
+          testNegativeFormErrors(documentFormErrorsNegative, errorTitle)
+          testDecimalFormErrors(documentFormErrorsNotWhole, errorTitle)
+          testOutOfMaxValFormErrors(documentFormErrorsOutOfRange, errorTitle)
+        }
+      }
+    }
+  }
```
Comment (ginnyamazed) [inline test/views/correctReturn/HowManyPackagedAsContractPackerViewSpec.scala:84]: page should also be verifying the detail info about a small producer

**Review by ginnyamazed — CHANGES_REQUESTED** [review]

File: conf/messages.en (hunk 3 above)
Comment (JakeReid2020) [reply conf/messages.en:323]: in returns this looks like it needs editing too

**Review by ginnyamazed — DISMISSED** [review]

**Review by ginnyamazed — APPROVED** [review]

**Pull request merged**